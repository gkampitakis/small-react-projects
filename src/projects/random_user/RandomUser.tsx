import React, { ReactElement, useState, useEffect, useCallback } from 'react';
import {
  FaEnvelopeOpen,
  FaUser,
  FaCalendarTimes,
  FaMap,
  FaPhone,
  FaLock
} from 'react-icons/fa';
import { useBodyStyles, useDocTitle, useFetch } from '../../hooks';
import './index.scss';

interface Person {
  image: string;
  name: string;
  email: string;
  age: string;
  street: string;
  phone: string;
  password: string;
}

const URL = 'https://randomuser.me/api/';

export default function RandomUser(): ReactElement {
  useDocTitle('Random User');
  useBodyStyles({
    background: 'hsl(210, 36%, 96%)'
  });
  const [title, setTitle] = useState('name');
  const [value, setValue] = useState('random person');
  const [person, setPerson] = useState<Person>({
    name: 'random person',
    image: 'https://randomuser.me/api/portraits/men/75.jpg',
    email: 'random@email.com',
    age: '26',
    street: 'Oxford Street 51',
    phone: '0332100000',
    password: '1234$%1'
  });
  const [data, loading, refetch, error] = useFetch<any>(URL);

  const getPerson = useCallback(() => {
    if (!data) return;
    if (error) return console.error(error);

    const result = data.results[0];
    const { phone, email } = result;
    const { large: image } = result.picture;
    const { password } = result.login;
    const { first, last } = result.name;
    const {
      dob: { age }
    } = result;
    const {
      street: { number, name }
    } = result.location;
    const newPerson = {
      image,
      phone,
      email,
      password,
      age,
      street: `${number} ${name}`,
      name: `${first} ${last}`
    };

    setTitle('name');
    setValue(newPerson.name);
    setPerson(newPerson);
  }, [data, error]);

  function handleValue(e: any) {
    if (e.target.classList.contains('icon')) {
      const newValue: keyof Person = e.target.dataset.label;
      setTitle(newValue);
      setValue(person[newValue]);
    }
  }

  useEffect(() => {
    getPerson();
  }, [getPerson]);

  return (
    <main className="random_user_project">
      <div className="block bck-black"></div>
      <div className="block">
        <div className="container">
          <img
            src={person.image}
            alt="Random User"
            className={`${loading && `loading`}`}
          />
          <p className="user-title">My {title} is</p>
          <p className="user-value">{value}</p>
          <div className="values-list">
            <button
              className="icon"
              data-label="name"
              onMouseOver={handleValue}
            >
              <FaUser />
            </button>
            <button
              className="icon"
              data-label="email"
              data-testid="email"
              onMouseOver={handleValue}
            >
              <FaEnvelopeOpen />
            </button>
            <button className="icon" data-label="age" onMouseOver={handleValue}>
              <FaCalendarTimes />
            </button>
            <button
              className="icon"
              data-label="street"
              onMouseOver={handleValue}
            >
              <FaMap />
            </button>
            <button
              className="icon"
              data-label="phone"
              onMouseOver={handleValue}
            >
              <FaPhone />
            </button>
            <button
              className="icon"
              data-label="password"
              onMouseOver={handleValue}
            >
              <FaLock />
            </button>
          </div>
          <button className="btn" data-testid="change-btn" onClick={refetch}>
            {loading ? 'loading...' : 'random user'}
          </button>
        </div>
      </div>
    </main>
  );
}
