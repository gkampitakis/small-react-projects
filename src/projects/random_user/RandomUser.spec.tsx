import React from 'react';
import { FetchMock } from 'jest-fetch-mock/types';
import { render, screen, act, waitFor, waitForDomChange } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import RandomUser from './RandomUser';


const fetchMock = (fetch as FetchMock);

describe('Random User', () => {
  beforeAll(fetchMock.enableMocks);
  beforeEach(fetchMock.resetMocks);

  it('Should render correctly', async () => {
    await act(async () => {
      const { baseElement } = render(<RandomUser />);

      expect(baseElement).toMatchSnapshot();
    });
  });

  it('Should change details displayed', async () => {
    fetchMock.mockResponseOnce(JSON.stringify({
      results: [{
        phone: 'mockPhone',
        email: 'mockEmail',
        picture: { large: 'mockImage' },
        login: { password: 'mockPassword' },
        name: {
          first: 'mockFirst',
          last: 'mockLast'
        },
        dob: {
          age: '26'
        },
        location: {
          street: {
            number: '123',
            name: 'mockStreet'
          }
        }
      }]
    }));

    const { baseElement } = render(<RandomUser />);

    const btn = screen.getByTestId('change-btn');

    await waitFor(() => btn);

    userEvent.hover(screen.getByTestId('email'));

    expect(baseElement).toMatchSnapshot();
  });
});
