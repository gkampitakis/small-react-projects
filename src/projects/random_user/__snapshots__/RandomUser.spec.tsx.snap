// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Random User Should change details displayed 1`] = `
<body
  style="background: rgb(241, 241, 241);"
>
  <div>
    <main
      class="random_user_project"
    >
      <div
        class="block bck-black"
      />
      <div
        class="block"
      >
        <div
          class="container"
        >
          <img
            alt="Random User"
            class="false"
            src="mockImage"
          />
          <p
            class="user-title"
          >
            My 
            email
             is
          </p>
          <p
            class="user-value"
          >
            mockEmail
          </p>
          <div
            class="values-list"
          >
            <button
              class="icon"
              data-label="name"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                />
              </svg>
            </button>
            <button
              class="icon"
              data-label="email"
              data-testid="email"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M512 464c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V200.724a48 48 0 0 1 18.387-37.776c24.913-19.529 45.501-35.365 164.2-121.511C199.412 29.17 232.797-.347 256 .003c23.198-.354 56.596 29.172 73.413 41.433 118.687 86.137 139.303 101.995 164.2 121.512A48 48 0 0 1 512 200.724V464zm-65.666-196.605c-2.563-3.728-7.7-4.595-11.339-1.907-22.845 16.873-55.462 40.705-105.582 77.079-16.825 12.266-50.21 41.781-73.413 41.43-23.211.344-56.559-29.143-73.413-41.43-50.114-36.37-82.734-60.204-105.582-77.079-3.639-2.688-8.776-1.821-11.339 1.907l-9.072 13.196a7.998 7.998 0 0 0 1.839 10.967c22.887 16.899 55.454 40.69 105.303 76.868 20.274 14.781 56.524 47.813 92.264 47.573 35.724.242 71.961-32.771 92.263-47.573 49.85-36.179 82.418-59.97 105.303-76.868a7.998 7.998 0 0 0 1.839-10.967l-9.071-13.196z"
                />
              </svg>
            </button>
            <button
              class="icon"
              data-label="age"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M436 160H12c-6.6 0-12-5.4-12-12v-36c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48v36c0 6.6-5.4 12-12 12zM12 192h424c6.6 0 12 5.4 12 12v260c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V204c0-6.6 5.4-12 12-12zm257.3 160l48.1-48.1c4.7-4.7 4.7-12.3 0-17l-28.3-28.3c-4.7-4.7-12.3-4.7-17 0L224 306.7l-48.1-48.1c-4.7-4.7-12.3-4.7-17 0l-28.3 28.3c-4.7 4.7-4.7 12.3 0 17l48.1 48.1-48.1 48.1c-4.7 4.7-4.7 12.3 0 17l28.3 28.3c4.7 4.7 12.3 4.7 17 0l48.1-48.1 48.1 48.1c4.7 4.7 12.3 4.7 17 0l28.3-28.3c4.7-4.7 4.7-12.3 0-17L269.3 352z"
                />
              </svg>
            </button>
            <button
              class="icon"
              data-label="street"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 117.66v346.32c0 11.32 11.43 19.06 21.94 14.86L160 416V32L20.12 87.95A32.006 32.006 0 0 0 0 117.66zM192 416l192 64V96L192 32v384zM554.06 33.16L416 96v384l139.88-55.95A31.996 31.996 0 0 0 576 394.34V48.02c0-11.32-11.43-19.06-21.94-14.86z"
                />
              </svg>
            </button>
            <button
              class="icon"
              data-label="phone"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M493.4 24.6l-104-24c-11.3-2.6-22.9 3.3-27.5 13.9l-48 112c-4.2 9.8-1.4 21.3 6.9 28l60.6 49.6c-36 76.7-98.9 140.5-177.2 177.2l-49.6-60.6c-6.8-8.3-18.2-11.1-28-6.9l-112 48C3.9 366.5-2 378.1.6 389.4l24 104C27.1 504.2 36.7 512 48 512c256.1 0 464-207.5 464-464 0-11.2-7.7-20.9-18.6-23.4z"
                />
              </svg>
            </button>
            <button
              class="icon"
              data-label="password"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M400 224h-24v-72C376 68.2 307.8 0 224 0S72 68.2 72 152v72H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48zm-104 0H152v-72c0-39.7 32.3-72 72-72s72 32.3 72 72v72z"
                />
              </svg>
            </button>
          </div>
          <button
            class="btn"
            data-testid="change-btn"
          >
            random user
          </button>
        </div>
      </div>
    </main>
  </div>
</body>
`;

exports[`Random User Should render correctly 1`] = `
<body>
  <div>
    <main
      class="random_user_project"
    >
      <div
        class="block bck-black"
      />
      <div
        class="block"
      >
        <div
          class="container"
        >
          <img
            alt="Random User"
            class="loading"
            src="https://randomuser.me/api/portraits/men/75.jpg"
          />
          <p
            class="user-title"
          >
            My 
            name
             is
          </p>
          <p
            class="user-value"
          >
            random person
          </p>
          <div
            class="values-list"
          >
            <button
              class="icon"
              data-label="name"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                />
              </svg>
            </button>
            <button
              class="icon"
              data-label="email"
              data-testid="email"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M512 464c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V200.724a48 48 0 0 1 18.387-37.776c24.913-19.529 45.501-35.365 164.2-121.511C199.412 29.17 232.797-.347 256 .003c23.198-.354 56.596 29.172 73.413 41.433 118.687 86.137 139.303 101.995 164.2 121.512A48 48 0 0 1 512 200.724V464zm-65.666-196.605c-2.563-3.728-7.7-4.595-11.339-1.907-22.845 16.873-55.462 40.705-105.582 77.079-16.825 12.266-50.21 41.781-73.413 41.43-23.211.344-56.559-29.143-73.413-41.43-50.114-36.37-82.734-60.204-105.582-77.079-3.639-2.688-8.776-1.821-11.339 1.907l-9.072 13.196a7.998 7.998 0 0 0 1.839 10.967c22.887 16.899 55.454 40.69 105.303 76.868 20.274 14.781 56.524 47.813 92.264 47.573 35.724.242 71.961-32.771 92.263-47.573 49.85-36.179 82.418-59.97 105.303-76.868a7.998 7.998 0 0 0 1.839-10.967l-9.071-13.196z"
                />
              </svg>
            </button>
            <button
              class="icon"
              data-label="age"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M436 160H12c-6.6 0-12-5.4-12-12v-36c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48v36c0 6.6-5.4 12-12 12zM12 192h424c6.6 0 12 5.4 12 12v260c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V204c0-6.6 5.4-12 12-12zm257.3 160l48.1-48.1c4.7-4.7 4.7-12.3 0-17l-28.3-28.3c-4.7-4.7-12.3-4.7-17 0L224 306.7l-48.1-48.1c-4.7-4.7-12.3-4.7-17 0l-28.3 28.3c-4.7 4.7-4.7 12.3 0 17l48.1 48.1-48.1 48.1c-4.7 4.7-4.7 12.3 0 17l28.3 28.3c4.7 4.7 12.3 4.7 17 0l48.1-48.1 48.1 48.1c4.7 4.7 12.3 4.7 17 0l28.3-28.3c4.7-4.7 4.7-12.3 0-17L269.3 352z"
                />
              </svg>
            </button>
            <button
              class="icon"
              data-label="street"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 117.66v346.32c0 11.32 11.43 19.06 21.94 14.86L160 416V32L20.12 87.95A32.006 32.006 0 0 0 0 117.66zM192 416l192 64V96L192 32v384zM554.06 33.16L416 96v384l139.88-55.95A31.996 31.996 0 0 0 576 394.34V48.02c0-11.32-11.43-19.06-21.94-14.86z"
                />
              </svg>
            </button>
            <button
              class="icon"
              data-label="phone"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M493.4 24.6l-104-24c-11.3-2.6-22.9 3.3-27.5 13.9l-48 112c-4.2 9.8-1.4 21.3 6.9 28l60.6 49.6c-36 76.7-98.9 140.5-177.2 177.2l-49.6-60.6c-6.8-8.3-18.2-11.1-28-6.9l-112 48C3.9 366.5-2 378.1.6 389.4l24 104C27.1 504.2 36.7 512 48 512c256.1 0 464-207.5 464-464 0-11.2-7.7-20.9-18.6-23.4z"
                />
              </svg>
            </button>
            <button
              class="icon"
              data-label="password"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M400 224h-24v-72C376 68.2 307.8 0 224 0S72 68.2 72 152v72H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48zm-104 0H152v-72c0-39.7 32.3-72 72-72s72 32.3 72 72v72z"
                />
              </svg>
            </button>
          </div>
          <button
            class="btn"
            data-testid="change-btn"
          >
            loading...
          </button>
        </div>
      </div>
    </main>
  </div>
</body>
`;
